packages/buidler-core/src/internal/buidler-evm/provider/filter.ts

```diff
@@ -0,0 +1,130 @@
+import Bloom from "@nomiclabs/ethereumjs-vm/dist/bloom";
+import { BN, bufferToHex, toBuffer } from "ethereumjs-util";
+
+import { RpcLogOutput } from "./output";
+
+export const LATEST_BLOCK = -1;
+
+export enum Type {
+  LOGS_SUBSCRIPTION = 0,
+  PENDING_TRANSACTION_SUBSCRIPTION = 1,
+  BLOCK_SUBSCRIPTION = 2
+}
+
+export interface FilterCriteria {
+  fromBlock: number;
+  toBlock: number;
+  addresses: Buffer[];
+  topics: Array<Array<Buffer | null> | null>;
+}
+
+export interface Filter {
+  id: number;
+  type: Type;
+  criteria?: FilterCriteria;
+  deadline: Date;
+  hashes: string[];
+  logs: RpcLogOutput[];
+  subscription: boolean;
+}
+
+export function bloomFilter(
+  bloom: Bloom,
+  addresses: Buffer[],
+  topics: Array<Array<Buffer | null> | null>
+): boolean {
+  if (addresses.length > 0 && !bloom.multiCheck(addresses)) {
+    return false;
+  }
+
+  for (const sub of topics) {
+    if (sub == null) {
+      continue;
+    }
+
+    let included = sub.length === 1;
+    for (const topic of sub) {
+      if (topic != null && bloom.check(topic)) {
+        included = true;
+        break;
+      }
+    }
+
+    if (!included) {
+      return false;
+    }
+  }
+  return true;
+}
+
+export function filterLogs(
+  logs: RpcLogOutput[],
+  criteria: FilterCriteria
+): RpcLogOutput[] {
+  const filteredLogs: RpcLogOutput[] = [];
+  for (const log of logs) {
+    const blockNumber = new BN(toBuffer(log.blockNumber!)).toNumber();
```

Comment:
I'd rather not cast `toNumber` here. It's improbable that this will overflow here, but if it happens it would probably be really hard to debug. For instance, `bufferToInt` is abused in parts of `ethereumjs`, and we constantly receive the same bug reports and have no idea what's going on :(

Reply:
Agree 👍